### PR TITLE
WebXR AR Improvements

### PR DESCRIPTION
**Pull request by slimbuck**
model-viewer-ar

This PR:
- makes improvements to AR mode:
    - improves light estimation by adding spherical harmonics
    - allow user to rotate the model in AR mode as well as place it
    - improve transition to AR
 - and in other news:
    - added wsad(qe) keyboard camera controls
    - fix grid clipping issue with tight near/far clip planes by clamping depth values
    - rewrote canvas/backbuffer resize logic
    - fix click-to-focus sometimes not working
    - correctly restore no skybox user option

### CODE REVIEW

File: src/read-depth.ts [hunk 1]
```diff
@@ -110,12 +99,15 @@ class ReadDepth {
             this.create();
         }
 
+        console.log(`tex=${this.texture.width}x${this.texture.height} depth=${depthTexture.width}x${depthTexture.height}`);
```
Comment (willeastcott) [inline src/read-depth.ts:102]: Did you mean to keep this in?

File: src/viewer.ts [hunk 2]
```diff
@@ -370,62 +372,122 @@ class Viewer {
 
     private initXrMode() {
         const backup = {
-            position: new Vec3()
+            position: new Vec3(),
+            rotation: new Vec3()
```
Comment (willeastcott) [inline src/viewer.ts:376]: I normally think of a rotation as a Quat and naming like `eulers` if you store Euler angles. Not a big deal tho.

File: src/viewer.ts [hunk 3]
```diff
@@ -370,62 +372,122 @@ class Viewer {
...
+                if (material.ambientSH !== sh) {
+                    material.ambientSH = sh;
+                    material.update();
+                }
+            });
         };
 
         const handlers = {
             starting: () => {
+                console.log('starting');
```
Comment (willeastcott) [inline src/viewer.ts:398]: Remove?

File: src/viewer.ts [hunk 4]
```diff
@@ -370,62 +372,122 @@ class Viewer {
...
+                this.setSkyboxBackground('None');
+                this.setSkyboxExposure(0);
+                this.setBackgroundColor(Color.BLACK);
+                this.app.scene.layers.getLayerById(LAYERID_SKYBOX).enabled = false;
+
+                this.sceneRoot.enabled = false;
             },
 
             started: () => {
+                console.log('started');
```
Comment (willeastcott) [inline src/viewer.ts:428]: Remove?

File: src/viewer.ts [hunk 5]
```diff
@@ -370,62 +372,122 @@ class Viewer {
...
                 this.app.scene.layers.getLayerById(LAYERID_SKYBOX).enabled = false;
-                this.app.scene.skyboxIntensity  = 0;
+                this.multiframe.blend = 0.5;
+                this.sceneRoot.enabled = true;
             },
 
-            place: (position: Vec3, rotation: Quat) => {
-                this.sceneRoot.enabled = true;
+            place: (position: Vec3) => {
+                console.log('place');
```
Comment (willeastcott) [inline src/viewer.ts:437]: Remove?

File: src/xr-mode.ts [hunk 6]
```diff
@@ -20,11 +20,99 @@ import {
...
+        this.touches.set(e.pointerId, {
+            previous: { x: e.clientX, y: e.clientY },
+            current: { x: e.clientX, y: e.clientY }
+        });
+    };
+
+    onPointerMove = (e: PointerEvent) => {
+        e.preventDefault();
+        e.stopPropagation();
+        console.log("xr pointermove");
```
Comment (willeastcott) [inline src/xr-mode.ts:54]: Remove?

File: src/xr-mode.ts [hunk 7]
```diff
@@ -20,11 +20,99 @@ import {
...
+            if (angle !== 0) {
+                this.onRotate?.(angle * -180 / Math.PI);
+            }
+        }
+    };
+
+    onPointerUp = (e: PointerEvent) => {
+        e.preventDefault();
+        e.stopPropagation();
+        console.log("xr pointerup");
```
Comment (willeastcott) [inline src/xr-mode.ts:82]: Remove?

**Comment by willeastcott** [comment]
Maybe add a vid/gif/screenshot or two to highlight the changes?

**Review by willeastcott — APPROVED** [review]
Really awesome improvements!!

File: src/read-depth.ts (hunk 1 above)
Comment (slimbuck) [reply src/read-depth.ts:102]: No! Thanks, removed.

File: src/viewer.ts (hunk 3 above)
Comment (slimbuck) [reply src/viewer.ts:396]: yep

File: src/viewer.ts (hunk 4 above)
Comment (slimbuck) [reply src/viewer.ts:426]: yep

File: src/xr-mode.ts (hunk 6 above)
Comment (slimbuck) [reply src/xr-mode.ts]: yep yep

File: src/viewer.ts (hunk 5 above)
Comment (slimbuck) [reply src/viewer.ts]: yep

File: src/xr-mode.ts (hunk 7 above)
Comment (slimbuck) [reply src/xr-mode.ts]: yep

**Pull request merged**